Add utility builders yocto

Mishytca opened the pull request:
Introduce internal utility dispatch and Yocto layer synchronization

This PR adds a generic internal --utility-* dispatch mechanism to Moulin. Utility invocations are detected before normal argument parsing and are forwarded to the corresponding builder-provided handler. The default build configuration flow remains unchanged when no utility option is used.

The Yocto builder now uses this mechanism to synchronize active BitBake layers with the layer list defined in the Moulin YAML manifest.

The Yocto synchronization utility:

- stores the list of Moulin-managed layers in a stamp/state file;
- synchronizes only Moulin-managed layers;
- leaves externally managed layers untouched;
- adds missing YAML layers;
- removes layers that were previously managed but disappeared from the YAML manifest;
- recreates managed layers when needed to preserve the YAML layer order.

Also added unit tests covering:

- Initial synchronization (no stamp file).
- External layers.
- Layer removal.
- Layer addition.
- Layer addition and removal.
- No-op synchronization.
- Layer reordering.
- Layer recreation to preserve YAML order.
- Shell quoting for layer paths containing spaces.

Documentation was updated to describe the internal --utility-* mechanism and clarify that these options are intended for internal use only.

Commit:
main.py: introduce generic utility dispatch

Moulin now detects utility invocations before the normal argument parsing,
dispatches execution to the corresponding utility handler, and skips the
standard build configuration flow.

The default behavior remains unchanged when no utility is specified.

No specific utilities are introduced in this commit; it only provides the
generic dispatch framework.

Signed-off-by: Mykhailo Androsiuk <mykhailo_androsiuk@epam.com>
Reviewed-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
main.py: introduce generic utility dispatch

Added a basic mechanism to run external utilities through --utility-*
arguments.
This allows Moulin to detect a utility flag, prepare the
configuration, and forward control to the corresponding handler.
Default behavior without utilities remains unchanged.
No specific utilities are included in this commit — only
the generic framework.

Signed-off-by: Mykhailo Androsiuk <mykhailo_androsiuk@epam.com>

Commit:
Add utility to sync Yocto layers with YAML layers, add unit tests

Added a utility that ensures the active Yocto build layers match
the configuration defined in YAML.
The synchronization process is tracked using a stamp file.

Utility workflow:
-- First run (does not stamp file):
	All specified layers from YAML are added with
	"bitbake-layers add-layer".
	A stamp file is created to record synchronization.

-- Subsequent runs (stamp exists):
	The current build layers are read via bitbake-layers show-layers.
	Paths are normalized, and Yocto core layers are excluded.
	The current set is compared with YAML-defined layers.
	Extra layers are removed, missing ones are added, and the stamp
	is updated.

Also added Unit tests are included to cover both scenarios:
-- Initial synchronization with no stamp file.
-- Updating layers when the stamp already exists
(removing unneeded layers and adding required ones).

Signed-off-by: Mykhailo Androsiuk <mykhailo_androsiuk@epam.com>

Commit:
Add utility to sync Yocto layers with YAML layers, add unit tests

Added a utility that ensures the active Yocto build layers match
the configuration defined in YAML.

The utility stores the list of layers managed by Moulin in a stamp
file. This allows it to synchronize only Moulin-managed layers while
leaving externally managed layers untouched.

Utility workflow:
- First run (stamp file does not exist):
  - Add all layers defined in the YAML manifest.
  - Record the managed layer list in the stamp file.

- Subsequent runs (stamp file exists):
  - Read the current layer list using "bitbake-layers show-layers".
  - Normalize layer paths.
  - Read the previously managed layer list from the stamp file.
  - Compare the managed layer set with the YAML manifest.
  - Remove layers that were previously managed but are no longer
    present in the YAML manifest.
  - Add layers that are missing from the current build.
  - If a simple remove/add operation would not preserve the YAML
    layer order, recreate the managed layers in the order defined
    by the YAML manifest.
  - Update the managed layer list stored in the stamp file.

Also add unit tests covering:
- Initial synchronization (no stamp file).
- Synchronization when only external layers are present.
- Synchronization when only layers need to be removed.
- Synchronization when only layers need to be added.
- Synchronization when layers need to be added and removed.
- No-op synchronization when layers already match.
- Layer reordering when the managed layer set matches but the
order differs.
- Synchronization when simple remove/add would not preserve YAML order.
- Shell quoting for layer paths containing spaces.

Signed-off-by: Mykhailo Androsiuk <mykhailo_androsiuk@epam.com>
Reviewed-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented on moulin/builders/yocto.py:
```diff
@@ -41,16 +46,22 @@ def gen_build_rules(generator: ninja_syntax.Writer):
-    # Add bitbake layers by calling bitbake-layers script
+    # Build a stable CLI that reuses the current Moulin invocation (script + args).
+    # This guarantees the utility gets the same build.yaml and global flags.
+    base_cli = shlex.join([os.path.basename(sys.argv[0]), *sys.argv[1:]])
```
I don't think that utility invocation needs all arguments, because it does not parse config file. So, `argv[0]` should be sufficient in this case

Mishytca replied:
Hi, @lorc,  I’ve fixed it.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,113 @@ def capture_state(self):
+        # Parse the table
+        lines = show.splitlines()
+
+        paths_only = []
+        start = None
+        for i, line in enumerate(lines):
+            if line.strip().startswith("layer") and "path" in line and "priority" in line:
+                start = i
+                break
+
+        # Output only table
+        if start is not None:
+            table_lines = lines[start:]
+            # Take the second column (paths)
+            for ln in table_lines[2:]:  # skip the header and "====="
+                s = ln.strip()
+                if not s or set(s) == {"="}:
+                    continue
+                parts = s.split()
+                if len(parts) >= 2:
+                    paths_only.append(parts[1])
+        else:
+            print("Couldn't find table in show-layers output")
```
I have a strong feeling that this part can be replaced by a simple regular expression. Also, it would be cool if you add example `bitbake show-layers` output in the comment above, so anyone trying to decipher this code will have some visual clue.

Mishytca replied:
Hi @lorc, done.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,113 @@ def capture_state(self):
+                     else (Path(args.yocto_dir) / work_dir_path)).resolve(strict=False)
+
+        # 3) Discard the first 3 layers (poky/meta, meta-poky, meta-yocto-bsp)
+        rest = canonical_paths[3:] if len(canonical_paths) > 3 else []
```
This is very wild assumption. What if first 3 layers will be different? Let me introduce you to meta-arago: https://layers.openembedded.org/layerindex/branch/master/layer/meta-arago-distro/ which is not based on meta-yocto-bsp as an example.

And why do you need to discard these layers in the first place?

Mishytca replied:
Hi @lorc.
Thank you for pointing this out.

You're right. The previous implementation relied on a Poky-specific assumption by treating a predefined set of layers as "core" layers. I have removed this assumption from the synchronization utility.

The utility now synchronizes only the layers managed by Moulin. It stores the list of managed layers in the stamp/state file and uses that information during subsequent synchronizations. As a result, externally managed layers remain untouched, regardless of the underlying Yocto distribution (Poky, meta-arago, or any other).

The existing Poky-specific filtering remains in the Yocto builder to preserve the previously implemented behavior for explicit Poky default layers, but the synchronization logic itself no longer depends on any Poky-specific layer classification.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,113 @@ def capture_state(self):
+            # 6) The quantity is the same --> compare the contents (taking into account the order)
+            desired_rel = [os.path.normpath(p) for p in args.layers]
+            current_rel = [os.path.normpath(p) for p in rel_paths]
+            if desired_rel != current_rel:
```
Is there a guarantee that `bitbake show-layers` will print layers in exact same order every time?

Mishytca replied:
Hi @lorc, done.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,113 @@ def capture_state(self):
+        """Remove current layers and add yaml layers; then touch the stamp."""
+        cmd = (
+            f"{env_prefix} && "
+            f"( bitbake-layers remove-layer {rel_paths_str} || true ) && "
```
You are referencing to a variable that was not declared prior to this. This is not good.

Mishytca replied:
Hi @lorc, done.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,113 @@ def capture_state(self):
+        """Remove current layers and add yaml layers; then touch the stamp."""
+        cmd = (
+            f"{env_prefix} && "
+            f"( bitbake-layers remove-layer {rel_paths_str} || true ) && "
```
Also, I think you should remove only layers that were removed by user

Mishytca replied:
Hi @lorc, done.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,113 @@ def capture_state(self):
+        cmd = (
+            f"{env_prefix} && "
+            f"( bitbake-layers remove-layer {rel_paths_str} || true ) && "
+            f"bitbake-layers add-layer {layers_str} && "
```
... and add layers that were added by user

Mishytca replied:
Hi @lorc, done.

lorc commented on tests/unittest/test_utility_builders_yocto_conf_missing.py:
```diff
@@ -0,0 +1,75 @@
+        show_cmd = mock_run.call_args[0][0]
+
+        # _run_bash called with 'bitbake-layers show-layers'
+        self.assertIn("bitbake-layers show-layers", show_cmd)
```
Unit test  should not test for implementation details. It is up to implementation how to get list of layers

Mishytca replied:
Hi @lorc, done.

lorc commented on tests/unittest/test_utility_builders_yocto_conf_missing.py:
```diff
@@ -0,0 +1,75 @@
+        # _run_bash called with 'bitbake-layers show-layers'
+        self.assertIn("bitbake-layers show-layers", show_cmd)
+
+        # cmd contains oe-init-build-env and work_dir
+        self.assertIn(f"cd {yocto_dir}", show_cmd)
+        self.assertIn("oe-init-build-env", show_cmd)
+
+        # sync_layers() -> check the compiled shell command
+        self.assertIsNotNone(mock_subproc.call_args)
+        bash_argv = mock_subproc.call_args[0][0]
+        self.assertEqual(bash_argv[:2], ["bash", "-lc"])
+        sync_cmd = bash_argv[2]
```
The same about implementation details

Mishytca replied:
Hi @lorc, done.

lorc commented on tests/unittest/test_utility_builders_yocto_conf_missing.py:
```diff
@@ -0,0 +1,75 @@
+        sync_cmd = bash_argv[2]
+
+        # remove empty list + add required layers + touch
+        self.assertIn("bitbake-layers remove-layer  || true", sync_cmd)
+        self.assertIn("bitbake-layers add-layer", sync_cmd)
+        for layer in layers:
+            self.assertIn(layer, sync_cmd)
+        self.assertIn(f"touch {stamp_path}", sync_cmd)
```
This is what test should check for: 
1. Unneeded layers were removed
2. Needed layers were added

Mishytca replied:
Hi @lorc, done.

lorc commented on tests/unittest/test_utility_builders_yocto_first_run.py:
```diff
@@ -0,0 +1,62 @@
+        self.assertNotIn("bitbake-layers show-layers", cmd)
+
+        # cmd no contains stamp_path
+        self.assertIn(f"touch {stamp_path}", cmd)
```
The same about implementation details. Test should check that correct list of layers was added and stamp was updated.

Mishytca replied:
Hi @lorc, done.

lorc reviewed (changes requested):
You should divide your patches differently:

1. First patch adds generic utility support 
2. Second patch adds yocto's sync layers functionality + tests 

Tests always should be added in the same commit as code that is being tested.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -41,16 +47,23 @@ def gen_build_rules(generator: ninja_syntax.Writer):
-    # Add bitbake layers by calling bitbake-layers script
+    # Minimal CLI: program + build.yaml (utility doesn't need global flags).
+    prog = os.path.basename(sys.argv[0])
+    conf_path = sys.argv[1]  # предполагаем, что это build.yaml
```
Could you please review your patch for yourself very carefully before submitting them?

Mishytca replied:
Hi @lorc, thank you for taking the time. I will recheck everything.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -41,16 +47,23 @@ def gen_build_rules(generator: ninja_syntax.Writer):
-    # Add bitbake layers by calling bitbake-layers script
+    # Minimal CLI: program + build.yaml (utility doesn't need global flags).
+    prog = os.path.basename(sys.argv[0])
+    conf_path = sys.argv[1]
```
You can't assume that path to config file will be in the first argument. See moulin usage:

`usage: moulin [-h] [--help-config] [-v] [--dump] build.yaml`

Why do you need it anyways?

Mishytca replied:
Hi, @lorc. You are absolutely right. I decided not to use the configuration in the utility.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -41,16 +47,23 @@ def gen_build_rules(generator: ninja_syntax.Writer):
+        "--yocto-dir $yocto_dir "
+        "--work-dir $work_dir "
+        "--layers $layers "
+        "--stamp $out",
```
I am not fully understand why you need `"  && ".join([])` part if you are basically providing single command. Care to elaborate?

Mishytca replied:
Hi @lorc. 
Using " && ".join([...]) is not allowed; the invocation must be a single command. Thanks for noticing this.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -41,16 +47,23 @@ def gen_build_rules(generator: ninja_syntax.Writer):
     generator.rule("yocto_add_layers",
                    command=f'bash -c "{cmd}"',
-                   description="Add yocto layers",
+                   description="Add and remove yocto layers",
```
Probably you need to rename rule to "yocto_update_layers" as well.

Mishytca replied:
Hi @lorc , done.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    """
+        Synchronize Yocto build layers with the specification from YAML.
+        The function ensures that the active Yocto layers in a build directory
+    match the layers defined in the YAML configuration. It uses a "stamp"
```
Strange formatting of the doc string

Mishytca replied:
Hi @lorc, done.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+         - Create the stamp file.
+
+      2. If the stamp file exists:
+         - Run 'bitbake-layers show-layers' to read the
```
to read the what?

Mishytca replied:
Hi, @lorc, done.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    # meta-python           /home/../../project/yocto/meta-openembedded/meta-python  5
+    # meta-networking       /home/../../project/yocto/meta-openembedded/meta-networking  5
+    # ...and so on
+    show_output = _run_bash(f"{env_prefix} && bitbake-layers show-layers", capture=True).stdout
```
I don't think this is a good variable name. You are not showing output to anyone

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    paths_only: List[str] = row_re.findall(show_output)
+
+    # Canonicalize build layers path
+    canonical_show_path = [str(Path(p).expanduser().resolve(strict=False)) for p in paths_only]
```
Again, why "show"?

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    parser.add_argument("--stamp", required=True)
+    args = parser.parse_args(argv)
+
+    # Extract arguments into local variables
```
I dunno why you need a bunch of local variables, some of them you use only once, but it is up to you...

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+        work_dir_path if work_dir_path.is_absolute()
+        else (Path(yocto_dir) / work_dir_path)).resolve(strict=False)
+
+    # CASE 1: No stamp file yet -> add layers and create stamp
```
Do you really need this special case? What is the difference with the next one?

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    actual_set = set(actual_layers_abs)
+
+    # Determine which layers should be removed
+    to_remove_abs = [p for p in actual_layers_abs if p not in spec_set]
```
`to_remove_abs = actual_set - spec_set`, no?

Mishytca replied:
Hi, @lorc. `to_add_abs` can indeed be viewed as spec - actual, but I intentionally keep it as a list comprehension to preserve the YAML order.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    # Determine which layers should be removed
+    to_remove_abs = [p for p in actual_layers_abs if p not in spec_set]
+    # Determine which layers should be added
+    to_add_abs = [p for p in spec_layers_abs if p not in actual_set]
```
Simillary,

`to_add_abs = spec_layers - actual_set`

Mishytca replied:
Hi @lorc . ` to_remove_abs` cannot be computed as actual - spec, because actual also contains layers that are not managed by Moulin. We only remove layers that were previously managed by Moulin and are no longer present in the YAML manifest. External layers must remain untouched.

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    # Filter out core layers to get only build-managed layers
+    used_build_layers_abs = [p for p in canonical_show_path if p not in yocto_core_layers_abs]
+
+    spec_layers_abs = yaml_layers_abs
```
Why you are introducing even more local variables? This function is already hard to follow...

Mishytca replied:
Hi @lorc, done

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    to_add_abs = [p for p in spec_layers_abs if p not in actual_set]
+
+    # If there are layers to remove, otherwise empty string
+    remove_cmd = ("bitbake-layers remove-layer " + " ".join(p for p in to_remove_abs)
```
isn't `p for p` looks a bit tautological for you? Why do you need loop in here?

Mishytca replied:
Hi @lorc, done

lorc commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +300,141 @@ def capture_state(self):
+    remove_cmd = ("bitbake-layers remove-layer " + " ".join(p for p in to_remove_abs)
+                  if to_remove_abs else "")
+    # If there are layers to add, otherwise empty string
+    add_cmd = ("bitbake-layers add-layer " + " ".join(p for p in to_add_abs)
```
The same tautological thing. 

Probably you wanted to write just `join(to_add_abs)` or `join(list(to_add_abs))` at worst.

Mishytca replied:
Hi @lorc, done

lorc reviewed (changes requested):
Please revisit use and naming of local variables. `handle_utility_call` is a total mess.

Mishytca commented:
> Please revisit use and naming of local variables. `handle_utility_call` is a total mess.
I reviewed the local variable names, tried to make the function simpler and more readable, and added several comments to make it clearer.

svlad-90 commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,135 @@ def capture_state(self):
+    cmd_parts = [env_prefix, remove_cmd, add_cmd, touch_cmd]
+    cmd = " && ".join(part for part in cmd_parts if part)
+
+    _run_bash(cmd)
```
Hi @Mishytca, will the "bitbake-layers show-layers" output after this command always match (including order) the layers listed in the moulin manifest YAML?

Mishytca replied:
Hello, @svlad-90! Thank you for your review and your comments.

I updated the logic to handle this case. Now Moulin checks whether a simple remove/add operation would preserve the YAML order. If not, it recreates the managed layers in the exact order defined by the YAML manifest.

svlad-90 commented on tests/unittest/test_utility_builders_yocto_sync_layers.py:
```diff
@@ -0,0 +1,124 @@
+import unittest
```
Hi @Mishytca,

The PR description lists four scenarios, but the tests cover only two: the first run without a stamp, and a run with a stamp that removes/adds. There is no separate test for “stamp exists, conf missing / only core layers”, no test for “identical layers -> no add/remove”, and no test for ordering.

Mishytca replied:
Thank you for pointing this out.

You're absolutely right. At the time of your review, those scenarios were indeed missing.

I've now extended the test coverage to include:

stamp exists with only Yocto core layers present;
stamp exists with identical layers (no add/remove expected);
layer reordering when the managed layer set matches but the order differs;
only add;
only remove;
content differs and requires layer recreation to preserve the YAML order;
shell quoting for layer paths containing spaces.
The PR description has also been updated.

svlad-90 commented on moulin/builders/yocto.py:
```diff
@@ -287,3 +298,135 @@ def capture_state(self):
         """
+
+
+def _env_prefix(yocto_dir: str, work_dir: str) -> str:
```
Hi @Mishytca,

Affected places: _env_prefix(), first-run bitbake-layers add-layer, remove_cmd, add_cmd, and touch_cmd.

The problem is that paths are inserted directly into the shell string. If yocto_dir, work_dir, a layer path, or the stamp path contains spaces or shell-sensitive characters, the command may break or execute something unintended.

Mishytca replied:
This has been addressed by quoting all paths before inserting them into the shell command. yocto_dir, work_dir, layer paths, and the stamp path are now wrapped with shlex.quote(), so paths containing spaces or other shell-sensitive characters are handled correctly.

svlad-90 commented on moulin/main.py:
```diff
@@ -30,6 +31,28 @@
+UTILITY_PREFIX = "--utility-"
+
+
+def _dispatch_utility(mod_qual: str, argv: List[str]) -> int:
```
Hi @Mishytca, please update the documentation regarding the utility mechanism.

Add it to this section:
https://moulin.readthedocs.io/en/latest/user-reference.html#internal-command-line-options

Mishytca replied:
The documentation has been updated.

Commit:
docs: user-reference: internal utility mechanism

Document the internal --utility-* command line mechanism and explain
how it is used by the Yocto builder.

Update the Yocto builder documentation to describe the current layer
synchronization behavior instead of the previous add-only workflo

Signed-off-by: Mykhailo Androsiuk <mykhailo_androsiuk@epam.com>
Reviewed-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc reviewed (approved):
`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`
